src/dataflow/src/logging/mod.rs

```diff
@@ -109,3 +121,89 @@ where
             .push(Event::Progress(vec![(self.time_ms as Timestamp, -1)]));
     }
 }
+
+/// A buffer that consolidates updates
+///
+/// The buffer is filled with updates until it reaches its capacity. At this point, the updates are
+/// consolidated to free up space. This process repeats until the consolidation recovered less than
+/// half of the buffer's capacity, at which point the buffer will be filled, once more consolidated
+/// and shipped.
+///
+/// The buffer retains a capability to send data on flush. It will flush all data once dropped, if
+/// time changes, or if the buffer capacity is reached.
+pub struct ConsolidateBuffer<'a, T, D: ExchangeData, R: Semigroup, P>
+where
+    P: Push<Bundle<T, (D, T, R)>> + 'a,
+    T: Clone + Lattice + Ord + TimelyTimestamp + 'a,
+    D: 'a,
+{
+    // a buffer for records, to send at self.cap
+    // Invariant: Buffer only contains data if cap is Some.
+    buffer: Vec<(D, T, R)>,
+    output_handle: OutputHandle<'a, T, (D, T, R), P>,
+    cap: Option<Capability<T>>,
+    clean: usize,
+    port: usize,
+}
+
+impl<'a, T, D: ExchangeData, R: Semigroup, P> ConsolidateBuffer<'a, T, D, R, P>
+where
+    T: Clone + Lattice + Ord + TimelyTimestamp + 'a,
+    P: Push<Bundle<T, (D, T, R)>> + 'a,
+{
+    /// Create a new [ConsolidateBuffer], wrapping the provided session.
+    ///
+    /// * `output_handle`: The output to send data to.
+    /// * 'port': The output port to retain capabilities for.
+    pub fn new(output_handle: OutputHandle<'a, T, (D, T, R), P>, port: usize) -> Self {
+        Self {
+            output_handle,
+            port,
+            cap: None,
+            buffer: Vec::with_capacity(Message::<T, (D, T, R)>::default_length()),
+            clean: 0,
+        }
+    }
+
+    /// Give an element to the buffer
+    pub fn give(&mut self, cap: &CapabilityRef<T>, data: (D, T, R)) {
+        // Retain a cap for the current time, which will be used on flush.
+        if self.cap.as_ref().map_or(true, |t| t.time() != cap.time()) {
+            // Flush on capability change
+            self.flush();
+            // Retain capability for the specified output port.
+            self.cap = Some(cap.delayed_for_output(cap.time(), self.port));
+        }
+        self.buffer.push(data);
+        if self.buffer.len() == self.buffer.capacity() {
+            // Consolidate while the consolidation frees at least half the buffer
+            if self.clean < self.buffer.capacity() / 2 {
+                consolidate_updates(&mut self.buffer);
+                self.clean = self.buffer.len();
+            } else {
+                // Consolidate once more and flush
+                consolidate_updates(&mut self.buffer);
+                self.flush();
+            }
```

Comment:
I think this logic may be slightly off, in that it seems possible that a consolidation happens, the buffer remains full, and is not flushed. A different way is, since you always call `consolidate_updates(&mut self.buffer);` is to just call that first, look at the results, and either ship or do not based on whether the buffer is at least half full.

Reply:
Thanks for pointing this out! I first implemented a scheme where the buffer[clean..] would be consolidated up until the buffer was full (or above a threshold), and then it would try to consolidate everything. But it's kind of hard to analyze its performance, so I reverted to a simpler approach. I updated the PR to remove the remains of `clean`.